Add command-line interface for `extra-labels`

nautofon opened the pull request:
@truckermudgeon

This'll be the 3rd step in the review of truckermudgeon#47.

nautofon committed:
Command-line interface for the extra-labels module

nautofon committed:
Code style: Avoid implicit type conversions in conditions

The only remaining cases of loose equality are null checks. These are not changed to strict equality in order to avoid having to check for undefined as well as null.

nautofon committed:
Add ats-towns submodule, update Makefile

nautofon commented:
Aside from the CLI tool, I'm adding the [ats-towns](https://github.com/nautofon/ats-towns) repo as a submodule here. Plus, I'm updating the Makefile to run this tool, creating the GeoJSON file for the demo map.

I don't know if the intermediate `usa-labels-meta.json` file is a solution you're really happy with. Personally I think it's good enough, but to avoid that file entirely, I suppose we could have the `extra-labels` module read CSV directly instead.

One important caveat: I can't actually run `make demo-data` because I don't have ETS2. I *think* my changes to the Makefile will work, but you'll want to verify that.

truckermudgeon commented on packages/clis/generator/commands/extra-labels.ts:
```diff
@@ -0,0 +1,191 @@
+    const metaDefaultsExisting = metaDefaults
+      .map(d => path.resolve(resourcesDir, d))
+      .filter(p => fs.existsSync(p));
+    metaDefaultsExisting.forEach(p => metaPaths.push(p));
```
I misread this at first, thinking that `metaDefaultsExisting`'s contents were being modified with the contents of `metaPaths`, which was already established as being empty.

Probably just a silly mistake on my part from skimming too fast. But I think:

```ts
metaPaths.push(...metaDefaultsExisting);
```

would make it clearer that `metaPaths` was the array being modified.

nautofon replied:
It's certainly more clear and shorter, too. Thanks!

truckermudgeon commented on packages/clis/generator/commands/extra-labels.ts:
```diff
@@ -0,0 +1,191 @@
+      analysis: label?.analysis,
+      result: label?.meta,
+    };
+    console.log(JSON.stringify(debug, null, 2));
```
`logger.log`? Or `logger.debug`, given comment above?

nautofon replied:
Changing it to `logger.log()`. The point of the `--token` option is to offer debug output without requiring the user to change the log level, so `logger.debug()` wouldn't give the desired behavior.

I see `logger.log()` and `console.log()` seem to work exactly the same. Is there a relevant difference between these two, other than consistency in the choice of object used for output?

truckermudgeon replied:
Consistency in the choice of object is the main thing. I'd like all console output to go through the Consola-backed `logger` object, so that, e.g., output can be controlled through its env variables, if I ever want to add timestamps to all console output I can do so easily, etc.

truckermudgeon commented on packages/clis/generator/commands/extra-labels.ts:
```diff
@@ -0,0 +1,191 @@
+}
+
+function cmp(a: string, b: string): number {
+  return a > b ? 1 : a < b ? -1 : 0;
```
Same thing as `return a.localeCompare(b)`?

nautofon replied:
I'm afraid it's not. `localeCompare()` does some locale-related magic hidden behind a bunch of [near-incomprehensible config options](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Collator/Collator#options). I'm looking for a good old simple ASCII comparison here.

Case in point: https://www.mycompiler.io/view/5EA2UdvhmFx

truckermudgeon replied:
Ahhh, got it. Good ol' ASCII comparison it is, then.

truckermudgeon commented on packages/clis/generator/geo-json/tests/extra-labels.test.ts:
```diff
@@ -435,3 +441,95 @@ describe('error checks', () => {
+  );
+  fs.writeFileSync(metaInPath, JSON.stringify([meta]));
+
+  const cli = yargs().command(extraLabels);
```
This setup should be done in a `beforeAll`, and the `test('clean up temp files')` should be turned into an `afterAll`.

The `test` blocks should be able to run individually, e.g., through the command line, or by pressing a button in an IDE:

<img width="420" height="160" alt="image" src="https://github.com/user-attachments/assets/067c0b60-658a-480c-a246-6b58aa626ca3" />

Currently, running an individual test inside the `describe('command-line interface')` scope will lead to temp files not being cleaned up 😕 .

I'd recommend something like:


```diff
diff --git a/packages/clis/generator/geo-json/tests/extra-labels.test.ts b/packages/clis/generator/geo-json/tests/extra-labels.test.ts
index 2e79a80..43aadc5 100644
--- a/packages/clis/generator/geo-json/tests/extra-labels.test.ts
+++ b/packages/clis/generator/geo-json/tests/extra-labels.test.ts
@@ -3,7 +3,7 @@ import fs from 'fs';
 import type { GeoJSON } from 'geojson';
 import os from 'os';
 import path from 'path';
-import { describe, expect, test, vi } from 'vitest';
+import { afterAll, describe, expect, test, vi } from 'vitest';
 import yargs from 'yargs';
 import * as extraLabels from '../../commands/extra-labels';
 import { logger } from '../../logger';
@@ -439,16 +439,24 @@ describe('error checks', () => {
     expect(loggerWarn).toHaveBeenCalledWith(
       "Can't assign metadata for target foobar unknown in usa",
     );
+    loggerWarn.mockRestore();
   });
 });
 
 describe('command-line interface', () => {
-  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'vitest-maps-'));
-  const metaInPath = path.join(tmpDir, 'meta.json');
-  const metaOutPath = path.join(tmpDir, 'extra-labels.json');
-  const geojsonPath = path.join(tmpDir, 'extra-labels.geojson');
+  let tmpDir: string;
+  let metaInPath: string;
+  let metaOutPath: string;
+  let geojsonPath: string;
+  let cli: yargs.Argv;
+
+  beforeAll(() => {
+    tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'vitest-maps-'));
+    metaInPath = path.join(tmpDir, 'meta.json');
+    metaOutPath = path.join(tmpDir, 'extra-labels.json');
+    geojsonPath = path.join(tmpDir, 'extra-labels.geojson');
+    cli = yargs().command(extraLabels);
 
-  console.log(tmpDir);
     const mileageTargets = [
       ...Array.from(fixtures.mt_stregis.values()),
       ...Array.from(fixtures.ca_napa.values()),
@@ -475,8 +483,12 @@ describe('command-line interface', () => {
       JSON.stringify(mileageTargets),
     );
     fs.writeFileSync(metaInPath, JSON.stringify([meta]));
+  });
 
-  const cli = yargs().command(extraLabels);
+  afterAll(() => {
+    fs.rmSync(tmpDir, { recursive: true });
+  });
 
   test('dry run on no output dir', () => {
     const loggerFail = vi.spyOn(logger, 'fail');
@@ -486,6 +498,7 @@ describe('command-line interface', () => {
     );
     expect(fs.existsSync(metaOutPath)).toBeFalsy();
     expect(fs.existsSync(geojsonPath)).toBeFalsy();
+    loggerFail.mockRestore();
   });
 
   test('metadata in, geojson out', () => {
@@ -527,9 +540,4 @@ describe('command-line interface', () => {
       country: 'US-MT',
     });
   });
-
-  test('clean up temp files', () => {
-    fs.rmSync(tmpDir, { recursive: true });
-    expect(fs.existsSync(tmpDir)).toBeFalsy();
-  });
 });
```

nautofon replied:
I didn't even realize you could run these subtests individually. 😲 Very cool!

Right, given that, `beforeAll()` and `afterAll()` should be used here.

truckermudgeon commented on packages/clis/generator/geo-json/tests/extra-labels.test.ts:
```diff
@@ -435,3 +441,95 @@ describe('error checks', () => {
+  const cli = yargs().command(extraLabels);
+
+  test('dry run on no output dir', () => {
+    const loggerFail = vi.spyOn(logger, 'fail');
```
Modifying a global object my spying on it should be undone when those modifications are no longer needed... quickest way is to call `loggerFail.mockRestore()` at the end of this `test`.

(It looks like upgrading to Vitest 3 will let us use [`using`](https://vitest.dev/api/vi.html#vi-spyon) so we don't need to explicitly call `.mockRestore()`, which is very nice. Something for the future!)

nautofon replied:
Ahyes.

I've gotten too used to such modifications being reverted automatically when the guard object goes out of scope.

truckermudgeon commented on packages/clis/generator/commands/extra-labels.ts:
```diff
@@ -0,0 +1,191 @@
+export const describe =
+  'Generates map labels GeoJSON from parser output and optional metadata';
+
+const metaDefaults = ['usa-labels-meta.json'];
```
Nit:

```suggestion
const metaDefaults = ['usa-labels-meta.json'] as const;
```

(without `as const`, consumers can still change the contents of `metaDefaults` at runtime 😕).

nautofon replied:
I tend to view mutable defaults as a good thing. But in this case, I don't think it matters either way. Adding `as const`.

truckermudgeon commented on packages/clis/generator/commands/extra-labels.ts:
```diff
@@ -0,0 +1,191 @@
+      target: label?.target,
+      analysis: label?.analysis,
+      result: label?.meta,
+    };
```
WDYT of:

```ts
    const label = labels.find(l => l.meta.token === args.token);
    if (label) {
      // The label might not be a TargetLabel. But this works for debug output.
      const targetLabel = label as TargetLabel;
      const debug = {
        target: targetLabel.target,
        analysis: targetLabel.analysis,
        result: targetLabel.meta,
      };
      console.log(JSON.stringify(debug, null, 2));
      logger.success(`wrote debug output for mileage target "${args.token}"`);
    } else {
      logger.error(`mileage target token "${args.token}" not found`);
    }
    return;
```

Asking because three things stuck out:
- we're casting `label`, the result of `.find()`, as something that is definitely defined, even though it could be `undefined`
- because `label` is cast as something that's defined, it's a little odd to see the `?.` operators in the `debug` declaration
- logging an empty object `{ } ` when `label` isn't defined seems unnecessary

nautofon replied:
Ahyes, that's better.

truckermudgeon commented on packages/clis/generator/commands/extra-labels.ts:
```diff
@@ -0,0 +1,191 @@
+  } else {
+    file = path.join(args.outputDir, `${outFile}.geojson`);
+
+    // For GeoJSON output, skip labels with incomplete data and 'unnamed' labels
```
🤔 is this comment correct? 

For a label to have incomplete data, it must be missing easting or southing data... what if it's missing its `text` field?

Also, what's an `unnamed` label? One whose `kind` field is equal to `'unnamed'`? If so, then should that check appear somewhere in the `.filter`?

nautofon replied:
The comment *does* correctly describe the intent here, but was evidently not worded well enough to be useful. How about this instead?

```
// For GeoJSON output, skip labels not considered "valid", which are those
// with certain attributes missing or with a `kind` attribute of "unnamed".
// Optionally (with --all), output everything we have coordinates for.
```

(The `kind == "unnamed"` check is a part of `isValid` in `Label`.)

truckermudgeon replied:
That comment LGTM 👍

truckermudgeon commented:
> I don't know if the intermediate `usa-labels-meta.json` file is a solution you're really happy with. Personally I think it's good enough

I do, too 👍 

>  I can't actually run `make demo-data` because I don't have ETS2. I _think_ my changes to the Makefile will work, but you'll want to verify that.

Verified!

nautofon committed:
Make metaDefaults array immutable

nautofon committed:
Clarify code and comments

nautofon committed:
Fix CLI tests to correctly run individual subtests

Vitest allows running specific subtests individually, e.g. by using the --testNamePattern option. This change uses beforeAll() and afterAll() to set up and tear down the temporary files correctly in such a scenario.

nautofon committed:
Restore logger.fail() after mocking with vi.spyOn()

nautofon reviewed (commented):
Thank you for your suggestions! I learn something new from you every time. ☺️

truckermudgeon reviewed (approved):
🚀
